File: python/src/threads/memory/conformance.py
"""The provider conformance suites: the same checks the built-in
providers run, for any third-party provider. Each suite takes a factory that makes a fresh
provider on a store and returns the failed checks by name; an empty tuple passes.

    assert await memory_suite(lambda store: local_memory().bind(store)) == ()
"""

from collections.abc import Awaitable, Callable

from threads.log import EventId, ThreadId
from threads.memory.guard import Binder, scoped_knowledge, scoped_memory
from threads.memory.protocol import KnowledgeProvider, MemoryProvider
from threads.memory.types import Provenance, Scope
from threads.result import Err, Ok
from threads.store import SqliteStore

A = Scope(tenant_id="tenant_a", agent="support", scope="user_1")
B = Scope(tenant_id="tenant_b", agent="support", scope="user_1")
_PROVENANCE = Provenance(
    thread_id=ThreadId("0192e000-0000-7000-8000-000000000001"),
    event_ids=(EventId("0192e000-0000-7000-8000-000000000002"),),
)


def _clock() -> int:
    return 0


async def memory_suite(
    make: Callable[[SqliteStore], Awaitable[MemoryProvider]], *, key_conflicts: bool = True
) -> tuple[str, ...]:
    """`key_conflicts`: the provider can tell the same key with other content (a capability
    flag; the loop never reuses a key with other content, so it is not a safety rule)."""
    opened = await SqliteStore.open()
    if isinstance(opened, Err):
        raise AssertionError(opened.error.message)
    store = opened.value
    try:
        provider = await make(store)
        a = scoped_memory(provider, Binder(store.bindings("memory"), A, _clock))
        b = scoped_memory(provider, Binder(store.bindings("memory"), B, _clock))
        failed: list[str] = []
        saved = await a.remember("the user prefers tabs", "user", _PROVENANCE, "k1")
        again = await a.remember("the user prefers tabs", "user", _PROVENANCE, "k1")
        if not isinstance(saved, Ok) or again != saved:
            failed.append("same key, same content is a no-op")
        clash = await a.remember("something else", "user", _PROVENANCE, "k1")
        if key_conflicts and not (isinstance(clash, Err) and clash.error.code == "invalid"):
            failed.append("same key, other content is invalid")
        hits = await a.recall("tabs", 5)
        if not (isinstance(hits, Ok) and [h.text for h in hits.value] == ["the user prefers tabs"]):
            failed.append("recall finds a saved record in its scope")
        raw = await provider.recall(B, "tabs", k=5)
        other = await b.recall("user prefers tabs", 5)
        if not (isinstance(raw, Ok) and not raw.value and other == Ok([])):
            failed.append("another tenant never recalls it")
        if isinstance(saved, Ok):
            gone = await a.forget(saved.value.id, "k2")
            twice = await a.forget(saved.value.id, "k2")
            after = await a.recall("tabs", 5)
            if not (isinstance(gone, Ok) and isinstance(twice, Ok) and after == Ok([])):
                failed.append("a forgotten record is never recalled")
        missing = await a.forget("no-such-id", "k3")
        if not (isinstance(missing, Err) and missing.error.code == "not_found"):
            failed.append("errors are typed values")
        return tuple(failed)
    finally:
        await store.close()


async def knowledge_suite(
    make: Callable[[SqliteStore], Awaitable[KnowledgeProvider]],
) -> tuple[str, ...]:
    opened = await SqliteStore.open()
    if isinstance(opened, Err):
        raise AssertionError(opened.error.message)
    store = opened.value
    try:
        provider = await make(store)
        a = scoped_knowledge(provider, Binder(store.bindings("knowledge"), A, _clock))
        b = scoped_knowledge(provider, Binder(store.bindings("knowledge"), B, _clock))
        failed: list[str] = []
        text = "Refunds take 5 days.\n\nShipping is free over 50 euros."
        v1 = await a.ingest("policy.md", "text/markdown", text.encode(), "policy.md", "k1")
        hits = await a.search("refunds", 5, None, None)
        first = hits.value[0] if isinstance(hits, Ok) and hits.value else None
        if not (isinstance(v1, Ok) and first is not None and first.version == v1.value.version):
            failed.append("a hit names its admitted version")
        elif text.encode()[first.span.start : first.span.end].decode() != first.text:
            failed.append("a hit's span is its exact bytes")
        if (await b.search("refunds", 5, None, None)) != Ok([]):
            failed.append("another tenant never retrieves it")
        v2 = await a.ingest("policy.md", "text/markdown", b"Refunds take 9 days.", None, "k2")
        now = await a.search("refunds", 5, None, None)
        if not (isinstance(v2, Ok) and isinstance(now, Ok) and "9 days" in now.value[0].text):
            failed.append("a refreshed source is searched at its new version")
        elif isinstance(v1, Ok):
            old = await a.search("refunds", 5, None, v1.value.revision)
            if not (isinstance(old, Ok) and old.value and "5 days" in old.value[0].text):
                failed.append("a search as of a revision sees that revision")
        removed = await a.remove("policy.md", "k3")
        if not (isinstance(removed, Ok) and (await a.search("refunds", 5, None, None)) == Ok([])):
            failed.append("a removed source is excluded from new searches")
        bad = await a.ingest("scan.pdf", "application/pdf", b"%PDF", None, "k4")
        if not (isinstance(bad, Err) and bad.error.code == "invalid"):
            failed.append("a parse failure is an explicit error")
        return tuple(failed)
    finally:
        await store.close()

File: typescript/packages/core/src/memory/conformance.ts
import { sha256Hex } from "../hash";
import { ThreadId } from "../log";
import { ok, type Result } from "../result";
import { within } from "../sandbox/remote/fence";
import type {
  Binding,
  KnowledgeProvider,
  MemoryProvider,
  MemoryRecord,
  Scope,
} from "./protocol";
import { KnowledgeHit, MemoryHit } from "./protocol";

// The provider conformance suites (F14.8): what every memory and
// knowledge provider must do, the built-ins included. Each takes the host's test function, so
// any runner works (bun:test, vitest, pytest's twin). A case fails by throwing.

export type Runner = {
  readonly test: (name: string, body: () => Promise<void>) => void;
};

const A: Scope = { tenant_id: "tenant-a", agent: "suite", scope: "user:1" };
const B: Scope = { tenant_id: "tenant-b", agent: "suite", scope: "user:1" };
const ERRORS = new Set([
  "unavailable",
  "timeout",
  "invalid",
  "scope_violation",
  "not_found",
]);
const encoder = new TextEncoder();

function check(ok: boolean, what: string): void {
  if (!ok) throw new Error(`provider conformance: ${what}`);
}

/** The value, or a failure naming the typed error; a malformed error fails the case too. */
function value<T>(
  result: Result<T, { readonly code: string; readonly message: string }>,
  what: string,
): T {
  if (result.ok) return result.value;
  check(
    ERRORS.has(result.error.code),
    `${what}: untyped error ${result.error.code}`,
  );
  throw new Error(
    `provider conformance: ${what}: ${result.error.code}: ${result.error.message}`,
  );
}

/** The host calls a provider under a live lease: its transport's fence passes. */
const LIVE = { fence: async () => ok(undefined) };
async function live<T>(call: () => Promise<T>): Promise<T> {
  const done = await within(LIVE, call);
  if (!done.ok) throw done.error.error;
  return done.value;
}

function liveMemory(p: MemoryProvider): MemoryProvider {
  return {
    ...p,
    remember: (...a) => live(() => p.remember(...a)),
    recall: (...a) => live(() => p.recall(...a)),
    forget: (...a) => live(() => p.forget(...a)),
  };
}

function liveKnowledge(p: KnowledgeProvider): KnowledgeProvider {
  return {
    ...p,
    ingest: (...a) => live(() => p.ingest(...a)),
    remove: (...a) => live(() => p.remove(...a)),
    search: (...a) => live(() => p.search(...a)),
    get: (...a) => live(() => p.get(...a)),
    revision: (...a) => live(() => p.revision(...a)),
  };
}

const binding = (n: number): Binding => ({
  namespace: "suite-namespace",
  record_id: `suite-record-${n}`,
});

function record(text: string, n: number): MemoryRecord {
  return {
    text,
    origin: "user",
    provenance: {
      thread_id: ThreadId.parse("0192a000-0000-7000-8000-000000000001"),
      event_ids: [],
    },
    binding: binding(n),
  };
}

/**
 * The memory suite. `make` returns a fresh, empty provider for each case. Providers that
 * declare `writeEffect: "idempotent"` also prove the keyed-write rules.
 */
export function memoryProviderSuite(
  name: string,
  make: () => Promise<MemoryProvider>,
  runner: Runner,
): void {
  const { test } = runner;
  test(`${name}: recall finds a remembered record and echoes its binding`, async () => {
    const p = liveMemory(await make());
    const ref = value(
      await p.remember(A, record("the deploy day is friday", 1), "k1"),
      "remember",
    );
    const hits = MemoryHit.array().parse(
      value(await p.recall(A, "deploy day", { k: 5 }), "recall"),
    );
    const hit = hits.find((h) => h.id === ref.id);
    check(hit !== undefined, "the remembered record is recalled");
    check(hit?.text === "the deploy day is friday", "the text is exact");
    check(hit?.origin === "user", "the origin is kept");
    check(
      hit?.binding.namespace === binding(1).namespace &&
        hit.binding.record_id === binding(1).record_id,
      "the binding is echoed unchanged",
    );
  });
  test(`${name}: another tenant's recall never returns the record`, async () => {
    const p = liveMemory(await make());
    value(
      await p.remember(A, record("tenant a secret launch code", 2), "k2"),
      "remember",
    );
    const hits = MemoryHit.array().parse(
      value(await p.recall(B, "secret launch code"), "recall"),
    );
    check(hits.length === 0, "scope B sees nothing of scope A");
  });
  test(`${name}: a forgotten record is not recalled`, async () => {
    const p = liveMemory(await make());
    const ref = value(
      await p.remember(A, record("the cat is named miso", 3), "k3"),
      "remember",
    );
    value(await p.forget(A, ref.id, "f3"), "forget");
    const hits = MemoryHit.array().parse(
      value(await p.recall(A, "cat named miso"), "recall"),
    );
    check(!hits.some((h) => h.id === ref.id), "forgotten stays forgotten");
  });
  test(`${name}: keyed writes (idempotent providers)`, async () => {
    const p = liveMemory(await make());
    if (p.writeEffect !== "idempotent") return;
    const first = value(
      await p.remember(A, record("tea over coffee", 4), "k4"),
      "remember",
    );
    const again = value(
      await p.remember(A, record("tea over coffee", 4), "k4"),
      "remember again",
    );
    check(first.id === again.id, "the same key and record is a no-op");
    const other = await p.remember(A, record("coffee over tea", 4), "k4");
    check(
      !other.ok && other.error.code === "invalid",
      "the same key with another record is invalid",
    );
    const hits = MemoryHit.array().parse(
      value(await p.recall(A, "tea coffee"), "recall"),
    );
    check(
      hits.filter((h) => h.id === first.id).length === 1,
      "one record, not two",
    );
  });
}

/** The knowledge suite: versioned hits with spans, revisions, tombstones, scope, typed errors. */
export function knowledgeProviderSuite(
  name: string,
  make: () => Promise<KnowledgeProvider>,
  runner: Runner,
): void {
  const { test } = runner;
  const source = (text: string, n: number) => {
    const content = encoder.encode(text);
    return {
      doc: {
        source_id: "guide.md",
        media_type: "text/markdown",
        content,
        binding: binding(n),
      },
      key: `guide.md@${sha256Hex(content)}`,
    };
  };
  test(`${name}: a hit cites an exact version and span of the admitted bytes`, async () => {
    const p = liveKnowledge(await make());
    const s = source("# Refunds\n\nRefunds take five business days.\n", 1);
    const v = value(await p.ingest(A, s.doc, s.key), "ingest");
    check(
      v.content_sha256 === sha256Hex(s.doc.content),
      "the version records the content hash",
    );
    const hits = KnowledgeHit.array().parse(
      value(await p.search(A, "refunds business days"), "search"),
    );
    const hit = hits[0];
    check(
      hit?.doc_id === v.doc_id && hit.version === v.version,
      "the hit names the version",
    );
    const span = new TextDecoder().decode(
      s.doc.content.subarray(hit?.span.start, hit?.span.end),
    );
    check(span === hit?.text, "the span is the excerpt's exact bytes");
    check(
      hit?.binding.record_id === binding(1).record_id,
      "the binding is echoed",
    );
    const doc = value(await p.get(A, v.doc_id, v.version), "get");
    check(
      doc.content_ref.sha256 === v.content_sha256,
      "get returns the admitted artifact",
    );
  });
  test(`${name}: an update is a new version; a search as of the old revision sees the old one`, async () => {
    const p = liveKnowledge(await make());
    const one = source("The office opens at nine.\n", 1);
    const v1 = value(await p.ingest(A, one.doc, one.key), "ingest v1");
    const two = source("The office opens at ten.\n", 2);
    const v2 = value(await p.ingest(A, two.doc, two.key), "ingest v2");
    check(
      v2.revision > v1.revision && v2.version !== v1.version,
      "a new version and revision",
    );
    const now = KnowledgeHit.array().parse(
      value(await p.search(A, "office opens"), "search"),
    );
    check(
      now.every((h) => h.version === v2.version),
      "a live search sees only the new version",
    );
    const then = KnowledgeHit.array().parse(
      value(await p.search(A, "office opens", { asOf: v1.revision }), "as of"),
    );
    check(
      then.some((h) => h.version === v1.version),
      "as of the old revision, the old version",
    );
  });
  test(`${name}: a removed source is excluded from new searches`, async () => {
    const p = liveKnowledge(await make());
    const s = source("Parking is on level two.\n", 1);
    value(await p.ingest(A, s.doc, s.key), "ingest");
    value(await p.remove(A, s.doc.source_id, "remove-1"), "remove");
    const hits = KnowledgeHit.array().parse(
      value(await p.search(A, "parking level"), "search"),
    );
    check(hits.length === 0, "removed means gone from new searches");
  });
  test(`${name}: another tenant's search never returns the document`, async () => {
    const p = liveKnowledge(await make());
    const s = source("The vault code is 4417.\n", 1);
    value(await p.ingest(A, s.doc, s.key), "ingest");
    const hits = KnowledgeHit.array().parse(
      value(await p.search(B, "vault code"), "search"),
    );
    check(hits.length === 0, "scope B sees nothing of scope A");
  });
  test(`${name}: failures are typed values, never throws`, async () => {
    const p = liveKnowledge(await make());
    const missing = await p.get(A, "nope.md", "1");
    check(
      !missing.ok && ERRORS.has(missing.error.code),
      "an unknown version is a typed error",
    );
  });
}
